contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1084,6 +1086,112 @@ tsql_coerce_string_literal_hook(ParseCallbackState *pcbstate, Oid targetTypeId,
 	return NULL;
 }
 
+static bool
+is_tsql_char_type(Oid type)
+{
+	common_utility_plugin *utilptr = common_utility_plugin_ptr;
+	return utilptr->is_tsql_bpchar_datatype(type) ||
+			utilptr->is_tsql_nchar_datatype(type);
+}
+
+static bool
+is_tsql_varchar_type(Oid type)
+{
+	common_utility_plugin *utilptr = common_utility_plugin_ptr;
+	return utilptr->is_tsql_varchar_datatype(type) ||
+			utilptr->is_tsql_nvarchar_datatype(type);
+}
+
+/* 
+ * Checks if the expression represents NULL or a string literal
+ */
+static bool
+expr_is_null_or_unknown(Node *expr)
+{
+	Oid type = exprType(expr);
+	return IsA(expr, Const) && (
+		type == UNKNOWNOID || type == TEXTOID || ((Const*)expr)->constisnull);
+}
+
+/* 
+ * When we must merge types together (i.e. UNION), if all types are
+ * null, literals, or [n][var]char types, then return the correct
+ * output type based on TSQL's precedence rules
+ * 
+ * If InvalidOid is returned, engine will handle finding a common type
+ * as usual
+ */
+static Oid
+tsql_select_common_type_hook(ParseState *pstate, List *exprs, const char *context,
```

Comment:
IMHO, just to be on safe side, we should call this hook only for UNION path as select_common_type is being used widely and we may break other operations ie, ISNULL.

Have we checked if there is any adverse impact on such operations due to this changes?

Reply:
Also, our test cases only revolve around set operations, whereas this changes may affect other operations as wel as I said, ISNULL, case-when, MERGE join, etc